### ISSUE TITLE
Decompiling multiple unreferenced assemblies into single syntax tree

### ISSUE THREAD

**Issue by Piedone** [icsharpcode/ILSpy#1834]
This is a **Question/Query** on using `ICSharpCode.Decompiler`. I hope this is the right place for this; if not, please point me where I should ask instead.

I've a project that uses old ILSpy libraries (specifically ICSharpCode.Decompiler v2.3.1 and ICSharpCode.NRefactory v5.5.1, and also Mono.Cecil v0.9.5.4) and am migrating it to the latest version.

How can you decompile multiple assemblies into a single AST with the current version, given the assemblies don't all reference each other? Previously this worked like this:

```
var assemblies = new List<AssemblyDefinition>();
...
var firstAssembly = assemblies.First();
var decompiledContext = new DecompilerContext(firstAssembly.MainModule);

var astBuilder = new AstBuilder(decompiledContext);
astBuilder.AddAssembly(firstAssembly);

foreach (var assembly in assemblies.Skip(1))
{
	astBuilder.AddAssembly(assembly);
}

var syntaxTree = astBuilder.SyntaxTree;
```
How would you do this today? I see there is a `TypeSystemAstBuilder` now which is kind of similarly initialized but then in my understanding isn't used as such. Rather, what I cobbled together based on the documentation, samples, and tests is this:

```
IEnumerable<string> assemblyPaths = ...

var firstAssemblyPath = assemblyPaths.First();
var firstModule = new PEFile(firstAssemblyPath, PEStreamOptions.PrefetchEntireImage);
var resolver = new UniversalAssemblyResolver(
	firstAssemblyPath,
	true,
	firstModule.Reader.DetectTargetFrameworkId(),
	PEStreamOptions.PrefetchMetadata);
	
// A few resolver.AddSearchDirectory() calls...

var decompilerSettings = new DecompilerSettings
{
	AnonymousMethods = false
};

var typeSystem = new DecompilerTypeSystem(firstModule, resolver, decompilerSettings);
var decompiler = new CSharpDecompiler(typeSystem, decompilerSettings);
var syntaxTree = decompiler.DecompileWholeModuleAsSingleFile();
```

I don't see an option to add more modules to a `DecompilerTypeSystem` instance (unless I inherit from it and do it myself somehow), nor add more `DecompilerTypeSystem`s to a `CSharpDecompiler`. I really need a single AST in the end though.

BTW the project in question is [Hastlayer](https://hastlayer.com/), a tool that converts .NET into hardware implementations for faster execution and smaller power consumption. See the related issue there: https://github.com/Lombiq/Hastlayer-SDK/issues/20

Also, if anybody here would like to offer some consulting on this migration project to help out when we get stuck on questions we'd be more than happy to pay for it.

**Comment by siegfriedpammer** [comment]
> I don't see an option to add more modules to a `DecompilerTypeSystem` instance (unless I inherit from it and do it myself somehow), nor add more `DecompilerTypeSystem`s to a `CSharpDecompiler`.

A DTS consists of all types from one main assembly and all its (transitive) references and it would not make sense to add multiple independent assemblies to one DTS, that's why it is immutable. You will have to create one DTS and `CSharpDecompiler` instance per assembly you want to decompile.

> I really need a single AST in the end though.

You could create two DTS and CSharpDecompiler instances, then decompile each assembly and then try to merge the ASTs by using `tree1.Members.AddRange(tree2.Members);`. If the code must be recompilable, you will have to move all using statements to the top of the syntax tree.

Note: I haven't tried this, but looking at the code, it should work.

Hope, this helps!

**Comment by Piedone** [comment]
Thank you for your reply! No need to produce recompilable code at the end so I've tried this. First I got and "System.ArgumentException: 'Node is already used in another tree.'" exception so ended up with this:

`syntaxTree.Members.AddRange(otherSyntaxTree.Members.Select(member => member.Detach()));`

I have several other questions related to changes between the v2 and v5 libraries. Should I post them here as issues? As mentioned we'd also be happy to pay for such consulting to you @siegfriedpammer, to other contributors or to some common budget if there's one.
